### DISCUSSION TITLE
Support for ID namespaces in Sphinx-Needs

### DISCUSSION THREAD

**Discussion by twodrops** [useblocks/sphinx-needs#1088] [General]
I'm exploring a way to introduce a namespace concept for IDs within Sphinx-Needs.

## Why do we need a namespace in Sphinx-Needs?

This aims to align with common software development practices that allows the creation and maintenance of human-readable IDs to uniquely identify elements.

Unlike database-centric systems, docs-as-code relies on human-readable IDs instead of GUIDs. However, longer human-readable IDs can be challenging to maintain, necessitating the division of IDs into namespaces and individual identifiers.

EDIT:
See detailed usecases [here.](https://github.com/useblocks/sphinx-needs/discussions/1088#discussioncomment-8047425)

### Component A
 
```
//Namespace defintion
namespace com.company.project.componentA{
    MyClass{
   }
 }
``` 
 
### Component B
 
```
//Namespace Import
import com.company.project.componentA;
 
OtherClass extends MyClass{
}
```

In the above example `Component A` defines a namespace and owns it. `Component B` utilizes this namespace to reference `Component A`

## Current state in Sphinx-Needs

Sphinx-Needs currently offers partial support for namespaces when using [needs_external_needs](https://sphinx-needs.readthedocs.io/en/latest/configuration.html#needs-external-needs). By employing the `id_prefix` configuration, it's possible to add a "namespace" to all imported IDs. Consequently, other needs can refer to IDs using this designated namespace, mirroring the software example provided earlier.

## Open Issues 

Several issues need resolution for complete namespace support within Sphinx-Needs:

### Namespace support is limited to `needs_external_needs`

Namespace support remains confined to `needs_external_needs`. Consequently, the possibility to create a prefix (namespace) is solely available when using this feature. For instance, during a `needs.json` export, there's no mechanism to prefix a namespace. In the end, external tools consuming these IDs might encounter conflicts or need to implement similar mechanisms to `needs_external_needs` to append a namespace to each ID.

### Ownership of namespace in `needs_external_needs`

Unlike the conventional concept where namespaces are defined and owned by the source (e.g., "ComponentA" defining and owning the namespace used by "ComponentB"), in `needs_external_needs`, the namespace is defined and owned by the importer (e.g., "ComponentB"). This results in the need for each component to redefine the namespace independently, deviating from the conventional "namespace" concept and resembling the Python "import - as" keyword alias concept. (Thanks @StephanKuempel for this hint.)
A potential workaround involves centrally defining a namespace variable in `conf.py` for "ComponentA" and utilizing it in the id_prefix of "ComponentB." However, this approach isn't entirely elegant when viewed from a namespace rather than an alias perspective.

### Namespace usage within the same project not possible

`needs_external_needs` facilitates referencing elements via a namespace from an external project but restricts such referencing within the confines of the same project. This implies that while a namespace can be employed to reference "ComponentA" from "ComponentB," it cannot be used reciprocally within "ComponentA" itself.
There exist scenarios where employing namespaces consistently to refer to elements, regardless of their origin within or outside the project, makes sense.

### Namespace support at project level only

Contrary to languages like C++ or Python that support namespaces at various levels (e.g., C++ at class level, Python at module level), using `needs_external_needs`, Sphinx-Needs only allows namespace emulation at the project level. This limitation might prompt the creation of multiple documentation projects solely to leverage the namespace feature.

@StephanKuempel @mb-emag @r-o-b-e-r-t-o  fyi

**Comment by danwos** [comment]
Thanks for the great write-up.

In my understanding, namespaces are on the textual level already possible, but not used inside Sphinx-Needs.
You can name a Sphinx-Needs object `com_company_project_componentA`, just `.` can't be used, as it separates the need_parts.

However, parts of this namespace are not automatically created based on the folder or file they are part of.
So a need ID `REQ_1` at location `/docs/components/a/security` is not getting a complete namespace ID like `my_project:components:a:security:REQ1`.

From the technical point I have no concerns using namespace and see the big benefit, but from the usability point of view I have some concerns and open questions:

### How does linking work?
What ID needs to be given to create a valid link?

* `REQ_1` if it is part of the same namespace, here `my_project:components:a:security`
* Always `my_project:components:a:security:REQ_1`?
* Partial `.a:security:REQ_1`?
* Relative `..:..:b:loginREQ_1`, using `..` as some kind of "folder-up"?

### Which information to take into account to create the namespace elements?
I can think of the following sphinx-related elements:

* `project_id` from `conf.py`
* folder name
* file name
* Headlines
* Sphinx-Needs ID
* Extra file in a folder, like `__needs__`, which defines the namespace to be used from here on.

### Restructuring the docs

What happens if:

* Folders and files get renamed or moved?
* Headlines are changing?
* Documentation projects are integrated into other projects, for instance, component-projects into system-wide documentation?

### Expected new features?

And most importantly, why spend this extra work? What is the benefit for the user?

* Addressing collections of Needs in filters and co? `:filter: id in my_project:components`.
* What else?

If we can answer these questions and keep the usability as good as it is, I would be very happy to support namespaces in one of the next releases.

**Reply by twodrops** [reply]
@danwos Thanks a lot for your quick and detailed reply.

Probably, we could separate namespace and usability requirements for namespace. Most parts you mention, I feel, are related to automatic creation of namespace taking into account other information. From my point of view, this could be addressed in the next step.

I see currently these possibilities of having a namespace concept in Sphinx-Needs. We could potentially derive some requirements for Sphinx-Needs based on these.

## Alternative 1: Using detailed identifier

This is a simple solution that can be implemented using the existing Sphinx-Needs features.

In this solution, a nomenclature for the identifier shall be defined which makes the need unique
across projects. For example, `<COMPANY-NAME>_<PROJECT-NAME>_<COMPONENT-NAME>_<need-type>_<RUNNING-NUMBER>`

The nomenclature shall be used in the definition of IDs and while referencing it.

The nomenclature shall ensure that the need is unique in any context.
When needs are exported to external tools, each need shall by default have the entire nomenclature
and remains unique also in this context.

An additional build or a `needs.json` based tool needs to be setup which does a verification across all projects in a context to ensure that there are no duplicate IDs.

### Example

```rst
   .. req-sw:: Need 001
      :id: COMPANY1_PROJECT1_COMP1_REQ_SW_001
```

This need can be referred **within the project** or **any other project** as:

```rst
   :need:`COMPANY1_PROJECT1_COMP1_REQ_SW_001`
```

### Pros

1. Easy to setup with existing Sphinx-Needs features

### Cons

1. Need an additional complete build or a `needs.json` based tool for duplicate verification.
2. All references (inter-project and intra-project) need to specific the complete identifier which can be potentially long.
3. Definition of ID needs to specify the full nomenclature (no namespace concept).

### Alternative 2: Using namespace and identifier

This solution makes use of existing Sphinx-Needs features, but needs additional configuration and minor tool enhancements.

In this solution, a simple nomenclature for the identifier shall be defined and an additional namespace for the project.
The identifier shall be short and simple and unique within the project.
The identifier together with namespace shall be unique across projects.

For example,

**Namespace:** `<COMPANY-NAME>_<PROJECT-NAME>_<COMPONENT-NAME>`

**Identifier:** `<NEED-TYPE>_<RUNNING-NUMBER>`

This enables definition of a needs identifier, without having to repeat a potentially long and redundant namespace
as part of each identifier (compared to "Alternative 1").

The need shall be referred only using the identifier within the project and using namespace and identifier from outside the project.

The namespace shall ensure that the need is unique in any context.

When needs are exported to external tools from Sphinx-Needs, each need shall by default have only the identifier.
An additional tool needs to be developed to modify the exported needs to include the namespace as part of
the identifier. Also the references to the needs within the project needs to be migrated. References
from outside the project shall already have the full namespace and identifier, and needs no migration (due to the use of `needs_external_needs` which forces a namespace/prefix).

An additional build  or a `needs.json` based tool needs to be setup which does a verification within all projects in a context to ensure that there are no duplicate IDs.

### Example

And example for the above solution shall be as follows:

```rst
   .. req-sw:: need 001
      :id: COMP1_REQ_SW_001
```	  

This need can be referred **within the project** as:

```rst
   :need:`COMP1_REQ_SW_001`
```	

and **from another project** as:

```rst
   :need:`COMPANY1_PROJECT1_COMP1_REQ_SW_001`
```   

where `COMPANY1_PROJECT1` is the namespace defined for the project centrally.  

### Pros

1. Easy to setup with existing Sphinx-Needs features.
2. Definition of ID needs to specify only the identifier (without the namespace) making it very user-friendly.
3. All references (inter-project) shall specify only the identifier.  Only intra-project references shall have the complete namespace and identifier.

### Cons

1. Need an additional complete build  or a `needs.json` based tool for duplicate verification.
2. Need an additional ID namespace exporter for external tools.

**Comment by StephanKuempel** [comment]
How about this:

- Add a global needs_namespace for the whole project to conf.py. (e.g. mycompany.myproject). All needs defined in the project will get this namespace applied.
- Like in software, have an additional directive .. needs_namespace :: componentB
  This namespace (concatenated to the global namespace) will be applied to all needs defined in the rst document (maybe also definable on higher levels for subdocuments (in 
  combination with .. toctree ::)).
- needs.json also contains namespace for each need as new attribute
- needs_external_needs imports also the namespaces (not in id but as additional attribute for each need)
- For referencing, like in software, you can either reference a need with namespace (e.g. as an extra option) or if not reference using the local namespace

Just my quick 2 cents.

**Reply by danwos** [reply]
I'm still struggling with the benefit of a namespace.
For what is it needed?

I see 2 scenarios here: 

a) Outside a sphinx project as a global, company/team-wide unique identifier
b) Inside a Sphinx project, for clustering needs and make them somehow referencable as a group.

We can do a, b, or both. However, a possible implementation and the need for certain directives and automatisms are different.

## a) Global identifiers outside the project.
Here it would be enough to have a project-specific identifier, which is used as "prefix" for the needs.
As @StephanKuempel mentioned, this can be configured in `conf.py` and used for exports.

Namespace+ID is just an additional attribute for the need object. Gets exported, but is not used internally.

## b) Inside a sphinx project.
Helper directives like `.. needs_namepspace::` or automatism based on folder/file/headline are needed.
Filter support for namespaces-parts needs to be implemented.
Need-ID contains the full namespace.

We need to define the use case, where exactly we want to use namespaces, and what they are good for.
Otherwise, we mix here solutions for different use cases and everybody has some other scenarios in mind :)

**Reply by twodrops** [reply]
@danwos 
I agree that this discussion misses a bit of the context. I tried my best to keep this information generic, but in the process missed making the usecases clear. Sorry for that.

The usecase is (a) and (b) above but not separate as you mentioned but together. The usecases I see are:

- It shall be possible to have IDs for needs which are unique, that is, within the same (sphinx) project or across (sphinx) projects.
- It shall be possible to refer to IDs within the same (sphinx) project or from another (sphinx or non-sphinx) project uniquely.
- It shall be possible to use/refer to IDs within external tools (imported from Sphinx-Needs) uniquely.
- It shall be possible to detect duplication of IDs within the same project or across projects.
- The definition and referencing of IDs shall be user-friendly (this means, no unnecessary typing of long namespace pattern as part of each ID)
- (IDE support) The IDE shall propose an ID based on the namespace defined.
- (IDE support) The IDE shall validate/lint IDs and report any error that may arise by not conforming to the defined nomenclature/pattern of the namespace.

**Reply by danwos** [reply]
These are all requirements for the global identifier use case only ("a" from my previous comment).

Let me try to specify the implementation:

## Technical concept (First draft)
Sphinx-Needs object shall get a new attribute: `global_id`.
`global_id` is based on a unique project identifier + the need object ID + an optional need part id.
`global_id` syntax is `<project_id>:<need-id>.<need-part>`.
We use `:` to not get confused with need-parts.

The globally unique project identifier is defined inside `conf.py` under `needs_project_id`.
Example: `my_project` or `PRJ_BOND_007`

Global Need ID Examples:
* `my_project:REQ_1`  : just the need id
* `PRJ_BOND_007:34511.3` : id + need_part

### New/Updated Features

There shall be a new option `need_projects`, which defines known projects and the location of their Sphinx-Needs data.
Similar to `needs_external_needs`, but may be a successor of it.

Example:
```python
needs_projects: {
 "my_project": {
    'base_url': 'http://mydocs/my_project',
    'json_url':  'http://mydocs/my_project/needs.json',
    'version': '1.0',
    'id_prefix': 'ext_',
    'css_class': 'external_link',
  },
"PRJ_BOND_007": {
    'base_url': 'http://007-docs/',
    'json_url':  'http://007-docs/needs.json',
    'version': '1.0',
    'id_prefix': 'ext_',
    'css_class': 'external_link',
  },
}
```

This can be used in existing directives/roles to reference Sphinx-Needs objects from outside the project.
```rst
.. needimport::
   :project: my_project
   
:need:`PRJ_BOND_007.34511.3`
```

This all leads to some rules/conclusions:
* value of `global_id` is calculated and can't be overwritten
* the `global_id` must not be used inside the Sphinx project to reference needs of the same project. Everything stays as it is.
* As long as "needs_project_id" is not part of `needs_projects`, the `global_id` is unique.
* We can only separate `projects`, there are no other namespace elements like "components", "teams" or whatever.

## Just my 2 cents
I don't see a big advantage in the above concept, as it just introduces some kind of a unique project_id, which is not necessarily related to Sphinx-Needs objects and could be used for other Sphinx elements as well.

It could be also handled outside Sphinx-Needs, by just writing a small script, which updates all IDs in a `needs.json` as a post process.

But maybe I missed something, and for sure it can be a starting point for full namespace support, including namespace parts based on folders, headlines, or whatever. Plus the internal support for using upper namespace-levels in filters and co, like `my_project:components:security` for all security related needs. Would maybe make the usage of `tags` or `sections` obsolete.

**Reply by twodrops** [reply]
> I don't see a big advantage in the above concept, as it just introduces some kind of a unique project_id, which is not necessarily related to Sphinx-Needs objects and could be used for other Sphinx elements as well.

@danwos  I guess you are right that the namespace definition is something that can be reused, for example, also for intersphinx. In the end, we need a proper concept if we need to replace requirements management tools like DOORS with Sphinx-Needs. In such database based tools, IDs are generated centrally and managed (often with a human-readable ID and a GUID with a mapping between them). For docs-as-code, we need to achieve the same with a single human-readable ID across several sphinx projects (when a multi-repo/project approach is followed). A unique ID is highly important for process conformance for unique referencing of requirements, requirements history tracking,  requirements exchange and so on.

I will go ahead with a needs.json migrator and `needs_external_needs` for now. Will update back if this is not enough.

**Comment by StephanKuempel** [comment]
I dont know if it is generally a good idea to couple the namespace into the needs.id as mandatory. 
I would rather see the namespace as an additional attribute to be able to reference to other (maybe imported) namespaces' needs.

In directives (eg.) needextend:
.. needextend:: my_need
   :links: req_1
   :namespace: othernamespace

For the :need: role however there could be a syntax like :need:`othernamespace.req_1`.

Validation however should always use namespace+id for uniqueness.

**Reply by danwos** [reply]
I agree, if the namespaced ID is important only outside of a project, in this case, it can be an additional attribute and does not affect the current implementation.

**Reply by r-o-b-e-r-t-o** [reply]
It would require to enhance the checking for uniqueness to consider the namespace (namespace+id together is the fully qualified need ID) so you don't need to trust the imported needs serializations. 
Also we would still have the problem of how you would be able to reference a need from another than the local namespace, which would require to have a namespace+id for a fully qualified reference in a link option or the need role.
I'm in favor of being able to use something like the `import as` semantic of Python, i.e. there is a global unique namespace (controlled by the component providing the needs) and components using these needs can define their own abbreviation for that namespace (which I call a namespace prefix). In json serializations of the needs a definition of the namespace prefix would be provided so consumers would be able to expand the namespace to the fully qualified ID but keep the json (similar to the rst code) concise and readable.

Example: 
Namespace: `PRJ_BOND_007`
Namespace Prefix: `bond`

Example fully qualified need ID: `bond:34511.3`
Example fully qualified need ID expanded: `PR_BOND_007_34511.3`

**Comment by frank-fzi** [comment]
Rather than implementing another proprietary namespace concept from scratch, how about employing a proven and widely used standard, such as Uniform Resource Names (URNs)? This is supported across varying tools and programming languages. For python, there is rdflib which supports URNs as follows:

```python
from rdflib import Namespace

COMP1 = Namespace("urn:COMPANY1:PROJECT1:COMP1:")
requirement = COMP1.REQ_SW_001
```

Example fully qualified need ID prefixed: `COMP1.REQ_SW_001` (as written by editors)
Example fully qualified need ID expanded: `urn:COMPANY1:PROJECT1:COMP1:REQ_SW_001` (as serialized for information exchange)

**Pros**

* employing a standard approach avoids pitfalls as discussed here
* a short (not necessarily globally unique) namespace prefix can be used to simplify editing while employing a globally unique namespace for serialization
* Namespaces and IDs can easily be interpreted outside Sphinx by employing standard URN supporting libraries
* IDs can be handled either as plain strings or as instances of URN that allows to distinguish namespace from local name

**Cons**

* requires URN implementation or introduces dependency on existing implementation
